### PR TITLE
Add DTS partition support

### PR DESCRIPTION
**Pull request by agross-oss**
This set of patches implements the DTS partitions support for defining flash partitions.

### CODE REVIEW

File: scripts/extract_dts_includes.py [hunk 1]
```diff
@@ -240,9 +240,12 @@ def extract_reg_prop(node_address, names, defs, def_label, div):
     address_cells = reduced[address]['props'].get('#address-cells', address_cells)
     size_cells = reduced[address]['props'].get('#size-cells', size_cells)
 
+  if post_label == None:
```
Comment (d3zd3z) [inline scripts/extract_dts_includes.py]: Either:
  if post_label is None:
or
  if not post_label:

File: scripts/extract_dts_includes.py [hunk 2]
```diff
@@ -397,14 +400,21 @@ def extract_single(node_address, yaml, prop, key, prefix, defs, def_label):
...
   return
 
-def extract_property(yaml, node_address, y_key, y_val, names, prefix, defs):
+def extract_property(node_compat, yaml, node_address, y_key, y_val, names, prefix, defs, label_override):
 
   node = reduced[node_address]
-  def_label = convert_string_to_label(get_compat(node)).upper()
-  def_label += '_' + node_address.split('@')[-1].upper()
+
+  if yaml[node_compat].get('base_label') != None:
```
Comment (d3zd3z) [inline scripts/extract_dts_includes.py]: ... is not None

**Review by d3zd3z — APPROVED** [review]
This looks good to me, I've also tested this works with mcuboot and applications, allowing the partitioning information to be removed from the bootloader itself.

File: scripts/extract_dts_includes.py [hunk 3]
```diff
@@ -397,14 +400,21 @@ def extract_single(node_address, yaml, prop, key, prefix, defs, def_label):
...
   return
 
-def extract_property(yaml, node_address, y_key, y_val, names, prefix, defs):
+def extract_property(node_compat, yaml, node_address, y_key, y_val, names, prefix, defs, label_override):
 
   node = reduced[node_address]
-  def_label = convert_string_to_label(get_compat(node)).upper()
-  def_label += '_' + node_address.split('@')[-1].upper()
+
+  if yaml[node_compat].get('base_label') is not None:
```
Comment (mbolivar) [inline scripts/extract_dts_includes.py]: Would `if 'base_label' in yaml[node_compat]` work? Might be a bit more idiomatic.

File: scripts/extract_dts_includes.py [hunk 4]
```diff
@@ -397,14 +400,21 @@ def extract_single(node_address, yaml, prop, key, prefix, defs, def_label):
...
+    def_label = yaml[node_compat].get('base_label')
+  else:
+    def_label = convert_string_to_label(node_compat.upper())
+    def_label += '_' + node_address.split('@')[-1].upper()
+
+  if label_override != None:
```
Comment (mbolivar) [inline scripts/extract_dts_includes.py]: PEP008: "Comparisons to singletons like None should always be done with is or is not , never the equality operators."

https://www.python.org/dev/peps/pep-0008/

File: scripts/extract_dts_includes.py [hunk 5]
```diff
@@ -415,24 +425,39 @@ def extract_property(yaml, node_address, y_key, y_val, names, prefix, defs):
...
+    extract_single(node_address, yaml,
                    reduced[node_address]['props'][y_key], y_key,
                    prefix, defs, def_label)
 
   return
 
-def extract_node_include_info(reduced, node_address, yaml, defs, structs):
-  node = reduced[node_address]
-  node_compat = get_compat(node)
+def extract_node_include_info(reduced, root_node_address, sub_node_address, yaml, defs, structs, y_sub):
```
Comment (mbolivar) [inline scripts/extract_dts_includes.py]: PEP 008: "Limit all lines to a maximum of 79 characters."

File: scripts/extract_dts_includes.py [hunk 6]
```diff
@@ -415,24 +425,39 @@ def extract_property(yaml, node_address, y_key, y_val, names, prefix, defs):
...
+  node = reduced[sub_node_address]
+  node_compat = get_compat(reduced[root_node_address])
+  label_override = None
 
   if not node_compat in yaml.keys():
     return {}, {}
 
-  y_node = yaml[node_compat]
+  if y_sub == None:
```
Comment (mbolivar) [inline scripts/extract_dts_includes.py]: is None

File: scripts/extract_dts_includes.py [hunk 7]
```diff
@@ -415,24 +425,39 @@ def extract_property(yaml, node_address, y_key, y_val, names, prefix, defs):
...
+    y_node = yaml[node_compat]
+  else:
+    y_node = y_sub
+
+  if yaml[node_compat].get('use-property-label', False) != False:
```
Comment (mbolivar) [inline scripts/extract_dts_includes.py]: PEP008: "Don't compare boolean values to True or False using == ."

(I.e., `if if yaml[node_compat].get('use-property-label', False):`.)

File: scripts/extract_dts_includes.py [hunk 8]
```diff
@@ -415,24 +425,39 @@ def extract_property(yaml, node_address, y_key, y_val, names, prefix, defs):
...
+    for yp in y_node['properties']:
+      if yp.get('label') != None:
```
Comment (mbolivar) [inline scripts/extract_dts_includes.py]: is not None

File: scripts/extract_dts_includes.py [hunk 9]
```diff
@@ -415,24 +425,39 @@ def extract_property(yaml, node_address, y_key, y_val, names, prefix, defs):
...
+        if node['props'].get('label') != None:
```
Comment (mbolivar) [inline scripts/extract_dts_includes.py]: is not None

File: scripts/extract_dts_includes.py [hunk 10]
```diff
@@ -415,24 +425,39 @@ def extract_property(yaml, node_address, y_key, y_val, names, prefix, defs):
...
+          label_override = convert_string_to_label(node['props']['label']).upper()
+          break
 
   # check to see if we need to process the properties
   for yp in y_node['properties']:
     for k,v in yp.items():
+      if 'properties' in v:
+        for c in reduced.keys():
```
Comment (mbolivar) [inline scripts/extract_dts_includes.py]: `.keys()` is not necessary; dictionary iterators are over keys by default.

```python
In [1]: d = {1:2, 3:4}

In [2]: for k in d:
   ...:     print(k)
   ...:     
1
3
```

File: misc/Kconfig [hunk 11]
```diff
@@ -15,6 +15,7 @@ config HAS_FLASH_LOAD_OFFSET
 	This option is selected by targets having a FLASH_LOAD_OFFSET
 	and FLASH_LOAD_SIZE.
 
+if !HAS_DTS
```
Comment (mbolivar) [inline misc/Kconfig]: Will this break existing targets which have DTS and want to use flash, but don't define partitions defined yet? If so, could merging this patch be delayed until we've had time to fix those up?

**Review by mbolivar — CHANGES_REQUESTED** [review]
Some Python code style nitpicks and a question about existing applications on the last patch.

By the way, you can run a PEP008 linter on Python code, similarly to checkpatch:

https://pypi.python.org/pypi/pep8

File: misc/Kconfig (hunk 11 above)
Comment (agross-linaro) [inline misc/Kconfig]: no it won't break those.  It defaults to using flash0 information.

Comment (mbolivar) [inline misc/Kconfig]: OK, thanks.

**Comment by d3zd3z** [comment]
While we're on PEP-8, we should probably fix the indentation to be 4 spaces. But, I would argue that should be a separate patch.

**Comment by agross-linaro** [comment]
yeah i was going to do a pep-8 blanket patch modification separately.

File: scripts/extract_dts_includes.py (hunk 3 above)
Comment (agross-linaro) [inline scripts/extract_dts_includes.py]: yeah that is clearer

**Comment by agross-linaro** [comment]
last commit fixed all the PEP-8 comments from Marti

**Review by mbolivar — COMMENTED** [review]
Thanks for touching those up. Looking forward to having partition support merged -- we need it to build multiple applications that can be chainloaded by mcuboot without resorting to hacks and duplicated code.

File: dts/arm/yaml/partition.yaml [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+      constraint: "fixed-partitions"
+
+  - partition\d+:
+      type: string
+      generation: define
+
+      properties:
+        - reg:
+            type: array
+            description: mmio register space
```
Comment (galak) [inline dts/arm/yaml/partition.yaml]: these seems wrong, while is mmio space, saying its register space seems a little misleading. (I assume we are talking about flashes that are MMIO direct access)

**Review by galak — CHANGES_REQUESTED** [review]
Can you add to commit message details about how we are setting CONFIG_FLASH_LOAD.

Also, not sure best place, but would be nice to see example or doc for what partition info looks like in dts.

**Comment by d3zd3z** [comment]
I have a partition map for k64f I've been using, should we include that somewhere in the docs? Also, fairly soon after this goes in, we'll be adding partitions to many of the devices.

**Comment by agross-linaro** [comment]
Addressed comments and fixed one issue with making the partitions.yaml file common.

**Comment by galak** [comment]
@d3zd3z can you provide a pointer to the k64 partition.  Would like to see as an example.

**Review by galak — APPROVED** [review]

**Pull request merged**